File: server/jellyfin.mjs
import { basename, normalize } from 'node:path'

export function jellyfinItemToMatch(item) {
  if (!item) return null
  const streams = item.MediaSources?.[0]?.MediaStreams || []
  const video = streams.find((stream) => stream.Type === 'Video') || {}
  const height = Number(video.Height) || 0
  return {
    itemId: item.Id || '',
    name: item.Name || 'Library item',
    path: item.Path || '',
    qualityLabel: height >= 2160 ? '4K' : height ? `${height}p` : '',
    width: Number(video.Width) || 0,
    height,
  }
}

async function jellyfinFetch(baseUrl, apiKey, path, options = {}) {
  const base = String(baseUrl || '').replace(/\/+$/, '')
  const token = String(apiKey || '').trim()
  if (!base || !token) throw new Error('Jellyfin URL and API key are required.')
  const response = await fetch(`${base}${path}`, {
    ...options,
    headers: {
      'X-Emby-Token': token,
      accept: 'application/json',
      ...(options.headers || {}),
    },
  })
  if (!response.ok) {
    const text = await response.text().catch(() => '')
    throw new Error(text || `Jellyfin request failed (${response.status})`)
  }
  return response.json()
}

export async function lookupJellyfinLibrary({ baseUrl, apiKey, imdbId, contentType, season, episode }) {
  const provider = `imdb.${String(imdbId || '').trim()}`
  if (!imdbId) return null

  if (contentType === 'series' && season && episode) {
    const seriesBody = await jellyfinFetch(
      baseUrl,
      apiKey,
      `/Items?Recursive=true&IncludeItemTypes=Series&AnyProviderIdEquals=${encodeURIComponent(provider)}&Fields=ProviderIds`,
    )
    const series = Array.isArray(seriesBody?.Items) ? seriesBody.Items[0] : null
    if (!series?.Id) return null
    const episodeBody = await jellyfinFetch(
      baseUrl,
      apiKey,
      `/Shows/${series.Id}/Episodes?Season=${season}&Fields=Path,MediaSources`,
    )
    const items = Array.isArray(episodeBody?.Items) ? episodeBody.Items : []
    const match = items.find((item) => Number(item.IndexNumber) === Number(episode))
    return jellyfinItemToMatch(match)
  }

  const includeType = contentType === 'series' ? 'Series' : 'Movie'
  const body = await jellyfinFetch(
    baseUrl,
    apiKey,
    `/Items?Recursive=true&IncludeItemTypes=${includeType}&AnyProviderIdEquals=${encodeURIComponent(provider)}&Fields=Path,MediaSources`,
  )
  const items = Array.isArray(body?.Items) ? body.Items : []
  return jellyfinItemToMatch(items[0])
}

export async function batchLookupJellyfinLibrary({ baseUrl, apiKey, items }) {
  const matches = {}
  const list = Array.isArray(items) ? items : []
  const concurrency = 6
  let index = 0

  async function worker() {
    while (index < list.length) {
      const current = list[index]
      index += 1
      if (!current?.key || !current.imdbId) continue
      try {
        matches[current.key] = await lookupJellyfinLibrary({
          baseUrl,
          apiKey,
          imdbId: current.imdbId,
          contentType: current.contentType || 'movie',
        })
      } catch {
        matches[current.key] = null
      }
    }
  }

  await Promise.all(Array.from({ length: Math.min(concurrency, list.length || 1) }, () => worker()))
  return matches
}

export async function lookupJellyfinSeasonEpisodes({ baseUrl, apiKey, imdbId, season }) {
  const provider = `imdb.${String(imdbId || '').trim()}`
  if (!imdbId || !season) return []

  const seriesBody = await jellyfinFetch(
    baseUrl,
    apiKey,
    `/Items?Recursive=true&IncludeItemTypes=Series&AnyProviderIdEquals=${encodeURIComponent(provider)}&Fields=ProviderIds`,
  )
  const series = Array.isArray(seriesBody?.Items) ? seriesBody.Items[0] : null
  if (!series?.Id) return []

  const episodeBody = await jellyfinFetch(
    baseUrl,
    apiKey,
    `/Shows/${series.Id}/Episodes?Season=${season}&Fields=Path,MediaSources,IndexNumber`,
  )
  const items = Array.isArray(episodeBody?.Items) ? episodeBody.Items : []
  return items
    .filter((item) => item?.Id && Number.isFinite(Number(item.IndexNumber)))
    .map((item) => ({
      episode: Number(item.IndexNumber),
      match: jellyfinItemToMatch(item),
    }))
}

export async function fetchJellyfinFavorites({ baseUrl, apiKey }) {
  const body = await jellyfinFetch(
    baseUrl,
    apiKey,
    '/Users/Me/Items?Recursive=true&IncludeItemTypes=Movie,Series&Filters=IsFavorite&Fields=ProviderIds,Name,Type,ProductionYear,ImageTags',
  )
  const items = Array.isArray(body?.Items) ? body.Items : []
  return items
    .map((item) => {
      const imdbId = item.ProviderIds?.imdb || item.ProviderIds?.Imdb
      if (!imdbId) return null
      return {
        id: String(imdbId),
        type: item.Type === 'Series' ? 'series' : 'movie',
        name: item.Name || 'Unknown',
        releaseInfo: item.ProductionYear ? String(item.ProductionYear) : undefined,
      }
    })
    .filter(Boolean)
}

export function jellyfinPathForJob(job, { pathMapFrom, pathMapTo }) {
  const target = normalize(job.targetPath || '')
  const from = normalize(pathMapFrom || '')
  const to = normalize(pathMapTo || '/movies')
  if (target === from) return to
  if (target.startsWith(`${from}/`)) return `${to}${target.slice(from.length)}`
  return target
}

export async function findJellyfinItemForJob(job, { baseUrl, apiKey, pathMapFrom, pathMapTo }) {
  const jellyfinPath = jellyfinPathForJob(job, { pathMapFrom, pathMapTo })
  const targetPath = normalize(job.targetPath || '')
  const basenameName = basename(job.targetPath || '')
  const body = await jellyfinFetch(baseUrl, apiKey, '/Items?Recursive=true&IncludeItemTypes=Movie,Episode&Fields=Path&Limit=10000')
  const items = Array.isArray(body?.Items) ? body.Items : []
  const exact = items.find((item) => item.Path === jellyfinPath || (targetPath && item.Path === targetPath))
  if (exact) return exact
  if (basenameName) {
    return items.find((item) => item.Path?.endsWith(`/${basenameName}`) || item.Path?.endsWith(basenameName)) || null
  }
  return null
}

export async function verifyJellyfinImport({ baseUrl, apiKey, targetPath, pathMapFrom, pathMapTo }) {
  const target = normalize(String(targetPath || ''))
  if (!target) return null
  const from = normalize(String(pathMapFrom || ''))
  const to = normalize(String(pathMapTo || '/movies'))
  let mapped = target
  if (target === from) mapped = to
  else if (target.startsWith(`${from}/`)) mapped = `${to}${target.slice(from.length)}`
  const body = await jellyfinFetch(baseUrl, apiKey, '/Items?Recursive=true&IncludeItemTypes=Movie,Episode&Fields=Path&Limit=10000')
  const items = Array.isArray(body?.Items) ? body.Items : []
  const basenameName = basename(target)
  const exact = items.find((item) => item.Path === mapped || item.Path === target)
  if (exact) return { itemId: exact.Id || '', path: exact.Path || '' }
  if (basenameName) {
    const fuzzy = items.find((item) => item.Path?.endsWith(`/${basenameName}`) || item.Path?.endsWith(basenameName))
    if (fuzzy) return { itemId: fuzzy.Id || '', path: fuzzy.Path || '' }
  }
  return null
}

export async function refreshJellyfin({ baseUrl, apiKey }) {
  const base = String(baseUrl || '').replace(/\/+$/, '')
  if (!base) return
  await jellyfinFetch(base, apiKey, '/Library/Refresh', { method: 'POST' })
}

export async function waitForJellyfinImport(job, config, { sleep, appendJobLog }) {
  for (let attempt = 0; attempt < 8; attempt += 1) {
    const item = await findJellyfinItemForJob(job, config).catch((error) => {
      if (appendJobLog) appendJobLog(job, 'jellyfin.verify.failed', { error: error.message })
      return null
    })
    if (item) return item
    await sleep(attempt < 2 ? 1000 : 2500)
  }
  return null
}

File: src-tauri/src/jellyfin.rs
#[derive(serde::Serialize)]
pub(crate) struct JellyfinInfo {
    name: String,
    version: String,
}

#[derive(serde::Serialize)]
pub(crate) struct JellyfinLibraryMatch {
    #[serde(rename = "itemId")]
    item_id: String,
    name: String,
    path: Option<String>,
    #[serde(rename = "qualityLabel")]
    quality_label: Option<String>,
    width: Option<i64>,
    height: Option<i64>,
}

#[tauri::command]
pub(crate) async fn lookup_jellyfin_library(
    base_url: String,
    api_key: String,
    imdb_id: String,
    content_type: String,
    season: Option<i32>,
    episode: Option<i32>,
) -> Result<Option<JellyfinLibraryMatch>, String> {
    let base_url = normalized_base_url(&base_url)?;
    let client = reqwest::Client::new();
    let token = api_key.trim();
    let provider = format!("imdb.{}", imdb_id.trim());

    if content_type == "series" {
        if let (Some(season), Some(episode)) = (season, episode) {
            let series_url = format!(
                "{base_url}/Items?Recursive=true&IncludeItemTypes=Series&AnyProviderIdEquals={provider}&Fields=ProviderIds",
            );
            let series_body = client
                .get(&series_url)
                .header("X-Emby-Token", token)
                .send()
                .await
                .map_err(|error| error.to_string())?
                .error_for_status()
                .map_err(|error| error.to_string())?
                .json::<serde_json::Value>()
                .await
                .map_err(|error| error.to_string())?;
            let series_id = series_body
                .get("Items")
                .and_then(|value| value.as_array())
                .and_then(|items| items.first())
                .and_then(|item| item.get("Id"))
                .and_then(|value| value.as_str());
            let Some(series_id) = series_id else {
                return Ok(None);
            };

            let episode_url = format!(
                "{base_url}/Shows/{series_id}/Episodes?Season={season}&Fields=Path,MediaSources&UserId="
            );
            let episode_body = client
                .get(&episode_url)
                .header("X-Emby-Token", token)
                .send()
                .await
                .map_err(|error| error.to_string())?
                .error_for_status()
                .map_err(|error| error.to_string())?
                .json::<serde_json::Value>()
                .await
                .map_err(|error| error.to_string())?;

            let matched = episode_body
                .get("Items")
                .and_then(|value| value.as_array())
                .and_then(|items| {
                    items.iter().find(|item| {
                        item.get("IndexNumber")
                            .and_then(|value| value.as_i64())
                            == Some(episode as i64)
                    })
                });

            return Ok(matched.map(|item| item_to_match(item)));
        }

        let series_url = format!(
            "{base_url}/Items?Recursive=true&IncludeItemTypes=Series&AnyProviderIdEquals={provider}&Fields=Path",
        );
        let body = client
            .get(&series_url)
            .header("X-Emby-Token", token)
            .send()
            .await
            .map_err(|error| error.to_string())?
            .error_for_status()
            .map_err(|error| error.to_string())?
            .json::<serde_json::Value>()
            .await
            .map_err(|error| error.to_string())?;
        return Ok(body
            .get("Items")
            .and_then(|value| value.as_array())
            .and_then(|items| items.first())
            .map(item_to_match));
    }

    let movie_url = format!(
        "{base_url}/Items?Recursive=true&IncludeItemTypes=Movie&AnyProviderIdEquals={provider}&Fields=Path,MediaSources",
    );
    let body = client
        .get(&movie_url)
        .header("X-Emby-Token", token)
        .send()
        .await
        .map_err(|error| error.to_string())?
        .error_for_status()
        .map_err(|error| error.to_string())?
        .json::<serde_json::Value>()
        .await
        .map_err(|error| error.to_string())?;

    Ok(body
        .get("Items")
        .and_then(|value| value.as_array())
        .and_then(|items| items.first())
        .map(item_to_match))
}

fn item_to_match(item: &serde_json::Value) -> JellyfinLibraryMatch {
    let (width, height) = video_dimensions(item);
    JellyfinLibraryMatch {
        item_id: item
            .get("Id")
            .and_then(|value| value.as_str())
            .unwrap_or_default()
            .to_string(),
        name: item
            .get("Name")
            .and_then(|value| value.as_str())
            .unwrap_or("Library item")
            .to_string(),
        path: item
            .get("Path")
            .and_then(|value| value.as_str())
            .map(ToString::to_string),
        quality_label: height.map(|value| {
            if value >= 2160 {
                "4K".to_string()
            } else {
                format!("{value}p")
            }
        }),
        width,
        height,
    }
}

fn video_dimensions(item: &serde_json::Value) -> (Option<i64>, Option<i64>) {
    let streams = item
        .get("MediaSources")
        .and_then(|value| value.as_array())
        .and_then(|sources| sources.first())
        .and_then(|source| source.get("MediaStreams"))
        .and_then(|value| value.as_array());

    if let Some(streams) = streams {
        for stream in streams {
            if stream.get("Type").and_then(|value| value.as_str()) == Some("Video") {
                return (
                    stream.get("Width").and_then(|value| value.as_i64()),
                    stream.get("Height").and_then(|value| value.as_i64()),
                );
            }
        }
    }
    (None, None)
}

#[tauri::command]
pub(crate) async fn test_jellyfin(
    base_url: String,
    api_key: String,
) -> Result<JellyfinInfo, String> {
    let base_url = normalized_base_url(&base_url)?;
    let body = reqwest::Client::new()
        .get(format!("{base_url}/System/Info"))
        .header("X-Emby-Token", api_key.trim())
        .send()
        .await
        .map_err(|error| error.to_string())?
        .error_for_status()
        .map_err(|error| error.to_string())?
        .json::<serde_json::Value>()
        .await
        .map_err(|error| error.to_string())?;

    Ok(JellyfinInfo {
        name: body
            .get("ServerName")
            .and_then(|value| value.as_str())
            .unwrap_or("Jellyfin")
            .to_string(),
        version: body
            .get("Version")
            .and_then(|value| value.as_str())
            .unwrap_or("unknown")
            .to_string(),
    })
}

#[tauri::command]
pub(crate) async fn authenticate_jellyfin(
    base_url: String,
    username: String,
    password: String,
) -> Result<String, String> {
    let base_url = normalized_base_url(&base_url)?;
    let body = serde_json::json!({
        "Username": username,
        "Pw": password,
    });
    let response = reqwest::Client::new()
        .post(format!("{base_url}/Users/AuthenticateByName"))
        .header(
            "Authorization",
            r#"MediaBrowser Client="Torfin", Device="Desktop", DeviceId="torfin", Version="1.0.0-beta""#,
        )
        .json(&body)
        .send()
        .await
        .map_err(|error| error.to_string())?;

    let status = response.status();
    let body = json_response(response, "Jellyfin auth").await?;
    if !status.is_success() {
        return Err(
            body_detail(&body).unwrap_or_else(|| format!("Jellyfin login failed with {status}"))
        );
    }

    body.get("AccessToken")
        .and_then(|value| value.as_str())
        .map(ToString::to_string)
        .ok_or_else(|| "Jellyfin did not return an access token.".to_string())
}

#[tauri::command]
pub(crate) async fn refresh_jellyfin_library(
    base_url: String,
    api_key: String,
) -> Result<(), String> {
    let base_url = normalized_base_url(&base_url)?;
    reqwest::Client::new()
        .post(format!("{base_url}/Library/Refresh"))
        .header("X-Emby-Token", api_key.trim())
        .send()
        .await
        .map_err(|error| error.to_string())?
        .error_for_status()
        .map_err(|error| error.to_string())?;
    Ok(())
}

fn normalized_base_url(value: &str) -> Result<String, String> {
    let trimmed = value.trim().trim_end_matches('/');
    if trimmed.is_empty() {
        return Err("Enter a server URL.".to_string());
    }
    let parsed = reqwest::Url::parse(trimmed).map_err(|error| error.to_string())?;
    if parsed.scheme() != "http" && parsed.scheme() != "https" {
        return Err("Server URLs must start with http:// or https://".to_string());
    }
    Ok(trimmed.to_string())
}

async fn json_response(
    response: reqwest::Response,
    context: &str,
) -> Result<serde_json::Value, String> {
    let status = response.status();
    let text = response.text().await.map_err(|error| error.to_string())?;
    serde_json::from_str::<serde_json::Value>(&text).map_err(|error| {
        let preview = text.chars().take(500).collect::<String>();
        if preview.trim().is_empty() {
            format!("{context} returned {status} with an empty body")
        } else {
            format!("{context} returned {status} with invalid JSON: {error}. Body: {preview}")
        }
    })
}

fn body_detail(body: &serde_json::Value) -> Option<String> {
    body.get("detail")
        .or_else(|| body.get("error"))
        .or_else(|| body.get("message"))
        .or_else(|| body.get("msg"))
        .and_then(|value| value.as_str())
        .map(ToString::to_string)
}
